From 9f87d28e67d480db225a1969709a8758de84cfb1 Mon Sep 17 00:00:00 2001
From: beviu <contact@beviu.com>
Date: Thu, 4 Dec 2025 13:33:55 +0100
Subject: [PATCH 14/17] Make LOG print to stderr by default

---
 src/core.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/core.h b/src/core.h
index e516653..6186c77 100644
--- a/src/core.h
+++ b/src/core.h
@@ -78,9 +78,9 @@ extern uint64_t dramsize;
 #define GIGA_PFN_MASK   (GIGAPAGE_MASK ^ UINT64_MAX)
 
 extern FILE *logfile;
-//#define LOG(...) fprintf(stderr, __VA_ARGS__)
+#define LOG(...) fprintf(stderr, __VA_ARGS__)
 //#define LOG(...)	fprintf(logfile, __VA_ARGS__)
-#define LOG(str, ...) while(0) {}
+//#define LOG(str, ...) while(0) {}
 
 //#define LOGPOLICY(...) fprintf(stderr, __VA_ARGS__)
 //#define LOGPOLICY(...)	fprintf(logfile, __VA_ARGS__)
-- 
2.52.0

